Add support for PermissionStatus.onchange in workers

RupinMittal opened the pull request:
#### 65f6f1007c12d8371600f83bef65b52101a447f3
<pre>
Add support for PermissionStatus.onchange in workers
<a href="https://bugs.webkit.org/show_bug.cgi?id=245090">https://bugs.webkit.org/show_bug.cgi?id=245090</a>

Reviewed by Chris Dumez and Sihui Liu.

The onchange event of the PermissionStatus object is supposed to
be supported in both window and worker contexts. A previous patch
added support for window contexts and this patch adds support for
dedicated, service, and shared workers.

Since WebPermissionController::permissionChanged is only ever called
on the main thread, and it calls PermissionObserver::stateChanged,
the PermissionObserver it uses should be a main thread object.
Before this patch, when there was only support for window contexts,
all PermissionStatus objects were main thread objects. But now,
PermissionStatus objects can be created on worker threads as well.
When this happens, an associated MainThreadPermissionObserver is created
on the main thread and there is a one-to-one mapping between the worker
thread PermissionStatus object and the MainThreadPermissionObserver.
Then, when WebPermissionController::permissionChanged finds a
PermissionObserver to call stateChanged on, that PermissionObserver is
either a main thread PermissionStatus object (in the window case) or a
main thread MainThreadPermissionObserver, whose stateChanged function
will jump to the relevant worker thread and invoke the stateChanged
function of the associated worker thread PermissionStatus object.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/permissions/service-worker-permission-status-onchange-event-expected.txt: Added.
* LayoutTests/http/tests/permissions/service-worker-permission-status-onchange-event.html: Added.
* LayoutTests/http/tests/permissions/service-worker-permission-status-onchange-event.js: Added.
(self.onmessage):
* LayoutTests/http/tests/permissions/shared-worker-permission-status-onchange-event-expected.txt: Added.
* LayoutTests/http/tests/permissions/shared-worker-permission-status-onchange-event.html: Added.
* LayoutTests/http/tests/permissions/shared-worker-permission-status-onchange-event.js: Added.
(port.onmessage):
(onconnect):
* LayoutTests/http/tests/permissions/worker-permission-status-onchange-event-expected.txt: Added.
* LayoutTests/http/tests/permissions/worker-permission-status-onchange-event.html: Added.
* LayoutTests/http/tests/permissions/worker-permission-status-onchange-event.js: Added.
(onmessage):
* LayoutTests/imported/w3c/web-platform-tests/permissions/permissions-cg.https-expected.txt:
* LayoutTests/platform/wk2/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp: Added.
(WebCore::MainThreadPermissionObserver::MainThreadPermissionObserver):
(WebCore::MainThreadPermissionObserver::~MainThreadPermissionObserver):
(WebCore::MainThreadPermissionObserver::stateChanged):
* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h: Copied from Source/WebCore/Modules/permissions/PermissionStatus.h.
* Source/WebCore/Modules/permissions/MainThreadPermissionObserverIdentifier.h: Copied from Source/WebCore/Modules/permissions/PermissionObserver.h.
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionObserver.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::allMainThreadPermissionObservers):
(WebCore::PermissionStatus::create):
(WebCore::PermissionStatus::PermissionStatus):
(WebCore::PermissionStatus::~PermissionStatus):
(WebCore::PermissionStatus::stateChanged):
* Source/WebCore/Modules/permissions/PermissionStatus.h:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::query):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
(WebKit::WebPermissionController::permissionChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::handleQueryPermission):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::setGeolocationPermission):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: <a href="https://commits.webkit.org/254490@main">https://commits.webkit.org/254490@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/deb563a6032d8060ed28d4b1e6cfeb15c70409af

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89205 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33766 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20022 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98521 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154840 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93213 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32272 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/27818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81564 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92990 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94852 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25633 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/76126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25563 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80498 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/80472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68538 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30050 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/14534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29774 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/15497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3149 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33221 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/76126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31924 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/34588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -31,48 +31,97 @@
+void PermissionStatus::createAssociatedMainThreadPermissionObserver(ScriptExecutionContext& context)
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
+    callOnMainThread([this, protectedThis = Ref { *this }, contextIdentifier = workerGlobalScope.identifier(), origin = m_origin.isolatedCopy()]() {
```
Capturing protectedThis here means that the PermissionStatus object could now get destroyed on the main thread, even if it is a worker thread object. This is not correct and can lead to crashes.

One way to avoid it would be to re-dispatch protectedThis back to the contextThread once we're done with it on the main thread.

youennf replied:
I am not sure it will actually work since the context thread might be stopped at the time we are trying to hop it back to it to do the unref.
Maybe the object that got ref and passed to main thread should not be an ActiveDOMObject. and PermissionStatus would keep a ref to it.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -31,48 +31,97 @@
+        return;
+    }
+
+    m_mainThreadPermissionObserverIdentifier = nextMainThreadPermissionObserverIdentifier();
```
Can this be initialized in createAssociatedMainThreadPermissionObserver() ? It looks a bit awkward here.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -31,48 +31,97 @@
     
-    // FIXME: Add support for workers. Currently, this only works for Window objects.
-    if (isMainThread())
+    m_mainThreadPermissionObserverIdentifier = 0;
```
Should not be needed.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -101,4 +150,32 @@ void PermissionStatus::eventListenersDidChange()
 }
 
+MainThreadPermissionObserver::MainThreadPermissionObserver(WeakPtr<PermissionObserver> workerPermissionObserver, ScriptExecutionContextIdentifier workerContextIdentifier, const PermissionDescriptor& workerDescriptor, const ClientOrigin& workerOrigin, PermissionQuerySource source, Page* page, PermissionState state)
+    : m_identifier(nextMainThreadPermissionObserverIdentifier())
```
So MainThreadPermissionObserver::m_identifier doesn't match PermissionStatus::m_mainThreadPermissionObserverIdentifier ?

RupinMittal replied:
Oh this is an issue. I'll fix this.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -101,4 +150,32 @@ void PermissionStatus::eventListenersDidChange()
+{
+    m_state = permissionState;
+
+    ScriptExecutionContext::postTaskTo(m_workerContextIdentifier, [this, &permissionState](auto&) {
```
What guarantees that `this` is still alive when the lambda runs? We should probably just capture `workerPermissionObserver = m_workerPermissionObserver`, not `this`

cdumez replied:
Also, again, you can never capture variables such as `permissionState` by reference in an async lambda.

RupinMittal replied:
I believe that there is no guarantee that `this` is alive when the lambda runs. `this` is held in the NeverDestroyed map as long as JavaScript keeps the associated `PermissionStatus` object alive. So yeah, it's probably best to not capture `this`. 

But `m_workerPermissionObserver` is a WeakPtr. It's not safe to capture a WeakPtr in a lambda that will be run on another thread right?

cdumez replied:
You can capture and pass WeakPtr to any thread. What you need to be careful about is on what thread you check if the WeakPtr is null or not (Needs to the be same same the object gets destroyed on).

RupinMittal replied:
Right, but in order to call PermissionObserver::stateChanged in the callback, we need to check if m_workerPermissionObserver is null or not--and this callback runs on the worker thread but the m_workerPermissionObserver WeakPtr was constructed on the main thread.

RupinMittal replied:
Ah, ok one solution is that I make PermissionObserver threadSafeRefCounted, then instead of having to check if the WeakPtr m_workerPermissionObserver is null in the callback, we can check it on the main thread, if not null, then we can Ref it, pass the Ref to the call back, then re-dispatch the ref back to main thread to be destroyed. Then, instead of storing a uniquePtr to the MainThreadPermissionObserver in the map, we can just store RefPtr's. How's that sound?

cdumez replied:
You are confused. m_workerPermissionObserver refers to a worker object, thus null-checking it and using it on the worker thread is the correct usage. I wouldn't have recommended this if it wasn't correct.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -78,6 +85,40 @@ class PermissionStatus final : public PermissionObserver, public ActiveDOMObject
     std::atomic<bool> m_hasChangeEventListener;
+    PermissionQuerySource m_source;
+    WeakPtr<Page> m_page;
+    uint64_t m_mainThreadPermissionObserverIdentifier;
```
`{ 0 }`

RupinMittal replied:
I made the identifier an optional.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -31,48 +31,97 @@
-    // FIXME: Add support for workers. Currently, this only works for Window objects.
-    if (isMainThread())
-        PermissionController::shared().removeObserver(*this);
+    callOnMainThread([identifier = m_mainThreadPermissionObserverIdentifier]() {
```
We should not dispatch to the main thread if m_mainThreadPermissionObserverIdentifier is 0.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -31,48 +31,97 @@
 WTF_MAKE_ISO_ALLOCATED_IMPL(PermissionStatus);
 
-Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+static uint64_t nextMainThreadPermissionObserverIdentifier()
```
Please use an ObjectIdentifier, we don't use untyped identifiers anymore.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -31,48 +31,97 @@
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
+    callOnMainThread([this, protectedThis = Ref { *this }, contextIdentifier = workerGlobalScope.identifier(), origin = m_origin.isolatedCopy()]() {
+        addMainThreadPermissionObserver(m_mainThreadPermissionObserverIdentifier, makeUnique<MainThreadPermissionObserver>(this, contextIdentifier, m_descriptor, origin, m_source, m_page.get(), m_state));
```
Looks like there is an extra space before `this`?

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -28,40 +28,47 @@
     const PermissionDescriptor& descriptor() const final { return m_descriptor; }
-    const ScriptExecutionContext* context() const final { return ActiveDOMObject::scriptExecutionContext(); }
+    PermissionQuerySource source() const final { return m_source; }
+    Page* page() const final { return m_page.get(); }
```
Something is weird with the design here. Seems awkward for a PermissionStatus to have a pointer to the Page, given that the Page is a main thread object and PermissionStatus can live on worker threads.

This m_page data member and page() getter are foot-guns. Or is m_page null when PermissionStatus is not a main thread object?

RupinMittal replied:
Oh that's a good point. The reason for this design is that in WebPermissionController::permissionChanged, we use `observer->page()`. And by ensuring that both PermissionStatus and MainThreadPermissionObserver have a page() method, we can keep the code in WebPermissionController::permissionChanged the same. The m_page data member held by PermissionStatus comes from Permissions::query() where the page is passed in as an argument to the PermissionStatus constructor. But that page is not null, it's the page that a worker context has (the one with an empty Chrome client and such). I can change this so that m_page is null when PermissionStatus is not a main thread object and then this would be fine right?

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -31,48 +31,97 @@
+
+static HashMap<uint64_t, std::unique_ptr<MainThreadPermissionObserver>>& allMainThreadPermissionObservers()
+{
+    static NeverDestroyed<HashMap<uint64_t, std::unique_ptr<MainThreadPermissionObserver>>> map;
```
Is this only used on the main thread? If so, we should use `MainThreadNeverDestroyed`

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -31,48 +31,97 @@
+    return map;
+}
+
+static void addMainThreadPermissionObserver(uint64_t identifier, std::unique_ptr<MainThreadPermissionObserver>&& observer)
```
Not sure this function is super helpful, why can't we just inline its body whenever it is used? It is one line.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -31,48 +31,97 @@
+    allMainThreadPermissionObservers().add(identifier, WTFMove(observer));
+}
+
+static void removeMainThreadPermissionObserver(uint64_t identifier)
```
ditto.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -78,6 +85,40 @@ class PermissionStatus final : public PermissionObserver, public ActiveDOMObject
+};
+
+class MainThreadPermissionObserver final : public PermissionObserver {
+    WTF_MAKE_NONCOPYABLE(MainThreadPermissionObserver); WTF_MAKE_FAST_ALLOCATED;
```
WTF_MAKE_FAST_ALLOCATED should probably be on the next line.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -78,6 +85,40 @@ class PermissionStatus final : public PermissionObserver, public ActiveDOMObject
+    uint64_t m_mainThreadPermissionObserverIdentifier;
+};
+
+class MainThreadPermissionObserver final : public PermissionObserver {
```
Should probably be in its own file?

szewai commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
```diff
@@ -0,0 +1,65 @@
+
+namespace WebCore {
+
+MainThreadPermissionObserver::MainThreadPermissionObserver(WeakPtr<PermissionObserver> workerPermissionObserver, ScriptExecutionContextIdentifier workerContextIdentifier, PermissionState workerState, const PermissionDescriptor& workerDescriptor, PermissionQuerySource workerSource, Page* workerPage, const ClientOrigin& workerOrigin)
```
`WeakPtr<PermissionObserver>&&`, `WeakPtr<Page>&&`, `clientOrigin&&`
I don't think you need "worker" prefix when all members are set from PermissionObserver in Worker.

cdumez replied:
We should not be passing WeakPtr<PermissionObserver> by value as this may cause unnecessary refcounting churn. Ideally, we'd pass a `WeakPtr<PermissionObserver>&&`.

szewai commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
```diff
@@ -0,0 +1,65 @@
+    , m_workerOrigin(workerOrigin)
+    , m_identifier(MainThreadPermissionObserverIdentifier::generate())
+{
+    PermissionController::shared().addObserver(*this);
```
`ASSERT(isMainThread())`

szewai commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
```diff
@@ -0,0 +1,65 @@
+
+MainThreadPermissionObserver::~MainThreadPermissionObserver()
+{
+    PermissionController::shared().removeObserver(*this);
```
`ASSERT(isMainThread())`

szewai commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
```diff
@@ -0,0 +1,65 @@
+    , m_workerSource(workerSource)
+    , m_workerPage(workerPage)
+    , m_workerOrigin(workerOrigin)
+    , m_identifier(MainThreadPermissionObserverIdentifier::generate())
```
`MainThreadPermissionObserverIdentifier::generateThreadSafe()`

szewai commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -0,0 +1,71 @@
+
+    using PermissionObserver::weakPtrFactory;
+    using PermissionObserver::WeakValueType;
+    using PermissionObserver::WeakPtrImplType;
```
Not needed.

szewai commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -0,0 +1,71 @@
+    PermissionState m_workerState;
+    PermissionDescriptor m_workerDescriptor;
+    PermissionQuerySource m_workerSource;
+    Page* m_workerPage;
```
Should store a WeakPtr instead of a raw pointer.

szewai commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,87 @@
+void PermissionStatus::createAssociatedMainThreadPermissionObserver(ScriptExecutionContext& context)
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
+    callOnMainThread([this, protectedThis = Ref { *this }, contextIdentifier = workerGlobalScope.identifier(), origin = m_origin.isolatedCopy()]() mutable {
```
Why protected this instead of weakThis and passing weakThis to main thread?
Maybe @cdumez knows: is it safe to create weakThis in constructor?

cdumez replied:
Do we really need to capture `this` and `protectedThis` here? Would it be sufficient to capture `WeakPtr { *this }`.
Seems that's all MainThreadPermissionObserver needs, unless I missed something. This would avoid deref'ing the PermissionStatus object on the wrong thread too.

szewai commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,87 @@
 
-    if (auto* document = dynamicDowncast<Document>(scriptExecutionContext()); !document->isFullyActive())
+    auto* document = dynamicDowncast<Document>(scriptExecutionContext());
+    if (document && !document->isFullyActive())
```
[Spec](https://w3c.github.io/permissions/#onchange-attribute) says:
`If this's relevant global object is a Window object, ... If document is null or document is not fully active, terminate this algorithm.`
Shouldn't we also return if scriptExecutionContext() is null?

szewai commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -28,40 +28,49 @@
 class ScriptExecutionContext;
 
-class PermissionStatus final : public PermissionObserver, public ActiveDOMObject, public RefCounted<PermissionStatus>, public EventTarget  {
+class PermissionStatus final : public PermissionObserver, public ActiveDOMObject, public ThreadSafeRefCounted<PermissionStatus>, public EventTarget  {
```
What is this change for?

RupinMittal replied:
I had this because I was calling ref / deref on a PermissionStatus object on different threads. But as you suggested in your earlier comment, I can use weakThis, and that should remove the need for PermissionStatus to be ThreadSafeRefCounted.

szewai commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -179,12 +179,15 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
-        PermissionController::shared().query(ClientOrigin { document.topOrigin().data(), WTFMove(originData) }, permissionDescriptor, document.page(), source, [contextIdentifier, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-            ScriptExecutionContext::postTaskTo(contextIdentifier, [promise = WTFMove(promise), permissionState, permissionDescriptor](auto& context) mutable {
+        PermissionController::shared().query(ClientOrigin { document.topOrigin().data(), WTFMove(originData) }, permissionDescriptor, document.page(), source, [contextIdentifier, permissionDescriptor, promise = WTFMove(promise), source, page = document.page()](auto permissionState) mutable {
+            ScriptExecutionContext::postTaskTo(contextIdentifier, [promise = WTFMove(promise), permissionState, permissionDescriptor, source, &page](auto& context) mutable {
```
Create a WeakPtr of page and pass it to worker thread.

cdumez replied:
`&page` is not safe here. I repeat, you cannot capture anything by reference in async lambdas.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -104,19 +104,7 @@ void WebPermissionController::permissionChanged(WebCore::PermissionName permissi
-            continue;
-
-        query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, page, *source, [observer = WeakPtr { observer }](auto newState) {
+        query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, observer.page(), observer.source(), [observer = WeakPtr { observer }](auto newState) {
```
If page() is null and source() is not ServiceWorker and SharedWorker, we might not need to query again.

RupinMittal replied:
I believe that if page() is null and source() is not ServiceWorker and SharedWorker, then query() will fail (there is an assertion that checks this) since this is never supposed to happen.

szewai replied:
Assert does not mean return; it means crash

cdumez replied:
It documents that something is impossible. Not sure we need a runtime check if something can't happen?

szewai replied:
The assert in WebPermissionController::query() holds when there is a null check for the page; since Rupin drops the check in this patch, there is no guarantee that the assert holds.

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -0,0 +1,71 @@
+
+    MainThreadPermissionObserverIdentifier identifier() const { return m_identifier; }
+
+    using PermissionObserver::weakPtrFactory;
```
Why are these needed?

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -0,0 +1,71 @@
+
+#include "ClientOrigin.h"
+#include "MainThreadPermissionObserverIdentifier.h"
+#include "Page.h"
```
Can this be forward-declared?

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,87 @@
+
+        allMainThreadPermissionObservers().add(mainThreadPermissionObserverIdentifier, WTFMove(mainThreadPermissionObserver));
+
+        ScriptExecutionContext::postTaskTo(contextIdentifier, [protectedThis = WTFMove(protectedThis), identifier = mainThreadPermissionObserverIdentifier](auto&) {
```
As Youenn mentioned earlier, this is not safe since postTaskTo() can fail if the worker thread is about to exit.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -28,40 +28,49 @@
 class ScriptExecutionContext;
 
-class PermissionStatus final : public PermissionObserver, public ActiveDOMObject, public RefCounted<PermissionStatus>, public EventTarget  {
+class PermissionStatus final : public PermissionObserver, public ActiveDOMObject, public ThreadSafeRefCounted<PermissionStatus>, public EventTarget  {
```
Hopefully we won't need this change. Seems risky to make to ref / deref this object on different threads.

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -179,12 +179,15 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
-                else
-                    promise.resolve(PermissionStatus::create(context, *permissionState, permissionDescriptor));
+                else {
+                    if (source != PermissionQuerySource::DedicatedWorker)
```
Why did we even capture the page in this case? Maybe we should capture a WeakPtr<Page> and make sure it is null when needed, upon capture.

cdumez replied:
I am also a bit unclear why we don't want to pass a Page to PermissionStatus when the source is something else than a dedicated worker. E.g. should a Source::Window have a Page?

RupinMittal replied:
Source::Window should have a Page. This is in the case of if source is one of the worker types. We don't want to pass a Page if the source is a Service or Shared worker. But maybe it is better to capture a WeakPtr instead.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -104,19 +104,7 @@ void WebPermissionController::permissionChanged(WebCore::PermissionName permissi
-            continue;
-
-        query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, page, *source, [observer = WeakPtr { observer }](auto newState) {
+        query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, observer.page(), observer.source(), [observer = WeakPtr { observer }](auto newState) {
```
Can we add an `ASSERT(isMainThread());` at the beginning of this function since the code seems to assume we're on the main thread.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,87 @@
     auto* origin = context.securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
     m_origin = ClientOrigin { context.topOrigin().data(), WTFMove(originData) };
     
```
Would be nice to add such an assertion:
ASSERT(isMainThread() || !page);

The code wouldn't be safe if this assertion didn't hold.

RupinMittal replied:
But if the source is a Dedicated worker, then we would not be on the main thread and we would have a page. So this assertion would be hit, even though there's nothing wrong no? I don't quite understand why the code is unsafe in that scenario.

cdumez replied:
You were constructing a `WeakPtr<Page>` from a `Page*`. This was only safe on the main thread since the Page gets destroyed on the main thread.

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserverIdentifier.h:
```diff
@@ -0,0 +1,35 @@
+namespace WebCore {
+
+enum MainThreadPermissionObserverIdentifierType { };
+using MainThreadPermissionObserverIdentifier = ObjectIdentifier<MainThreadPermissionObserverIdentifierType >;
```
nit: extra space before `>;`

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -46,7 +46,7 @@ class PermissionController : public ThreadSafeRefCounted<PermissionController> {
     
     virtual ~PermissionController() = default;
-    virtual void query(ClientOrigin&&, PermissionDescriptor, Page*, PermissionQuerySource, CompletionHandler<void(std::optional<PermissionState>)>&&) = 0;
+    virtual void query(ClientOrigin&&, PermissionDescriptor, WeakPtr<Page>, PermissionQuerySource, CompletionHandler<void(std::optional<PermissionState>)>&&) = 0;
```
Please don't pass `WeakPtr<Page>` by value as this causes refcounting churn. `const WeakPtr<Page>&` is probably what we want here.

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -59,7 +59,7 @@ class DummyPermissionController final : public PermissionController {
 private:
     DummyPermissionController() = default;
-    void query(ClientOrigin&&, PermissionDescriptor, Page*, PermissionQuerySource, CompletionHandler<void(std::optional<PermissionState>)>&& callback) final { callback({ }); }
+    void query(ClientOrigin&&, PermissionDescriptor, WeakPtr<Page>, PermissionQuerySource, CompletionHandler<void(std::optional<PermissionState>)>&& callback) final { callback({ }); }
```
ditto.

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -0,0 +1,68 @@
+    const ClientOrigin& origin() const final { return m_origin; }
+    const PermissionDescriptor& descriptor() const final { return m_descriptor; }
+    PermissionQuerySource source() const final { return m_source; }
+    WeakPtr<Page> page() const final { return m_page; }
```
I think this should return a `Page*` or a `const WeakPtr<Page>&` to avoid refcounting churn. I think this function gets called off the main thread so you likely want `const WeakPtr<Page>&`

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
+    return map;
+}
+
+Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, WeakPtr<Page> page)
```
`WeakPtr<Page>&&`

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
+Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, WeakPtr<Page> page)
 {
-    auto status = adoptRef(*new PermissionStatus(context, state, descriptor));
+    auto status = adoptRef(*new PermissionStatus(context, state, descriptor, source, page));
```
`WTFMove(page)`

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
 }
 
-PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, WeakPtr<Page> page)
```
`WeakPtr<Page>&&` or `const WeakPtr<Page>&`

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
+void PermissionStatus::createAssociatedMainThreadPermissionObserver(ScriptExecutionContext& context)
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
+    callOnMainThread([this, weakThis = WeakPtr { *this }, contextIdentifier = workerGlobalScope.identifier(), origin = m_origin.isolatedCopy()]() mutable {
```
What is `this` used for?

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
 
+void PermissionStatus::createAssociatedMainThreadPermissionObserver(ScriptExecutionContext& context)
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
```
Why do you need to cast? Seems you should be able to call `identifier()` on a `ScriptExecutionContext`.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
+    callOnMainThread([this, weakThis = WeakPtr { *this }, contextIdentifier = workerGlobalScope.identifier(), origin = m_origin.isolatedCopy()]() mutable {
+        if (!weakThis)
```
This check is kinda meaningless. `this` gets destroyed on a worker thread so checking if weakThis is null or not on the main thread is super racy.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
+        if (!weakThis)
+            return;
+
+        auto mainThreadPermissionObserver = makeUnique<MainThreadPermissionObserver>(weakThis, contextIdentifier, m_state, m_descriptor, m_source, WeakPtr { m_page }, WTFMove(origin));
```
`this` is an object that gets destroyed on a worker thread and you are on the main thread here. It is not safe to access all these `m_*` variables here. We want to stop capturing `this` in the lambda and capture the `m_*` variables in the lambda instead.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
+
+        allMainThreadPermissionObservers().add(mainThreadPermissionObserverIdentifier, WTFMove(mainThreadPermissionObserver));
+
+        ScriptExecutionContext::postTaskTo(contextIdentifier, [weakThis, identifier = mainThreadPermissionObserverIdentifier](auto&) {
```
`weakThis = WTFMove(weakThis)`

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
+        allMainThreadPermissionObservers().add(mainThreadPermissionObserverIdentifier, WTFMove(mainThreadPermissionObserver));
+
+        ScriptExecutionContext::postTaskTo(contextIdentifier, [weakThis, identifier = mainThreadPermissionObserverIdentifier](auto&) {
+            if (weakThis)
```
Here checking weakThis is indeed safe since we are on its context thread.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
-    // FIXME: Add support for workers. Currently, this only works for Window objects.
-    if (isMainThread())
-        PermissionController::shared().removeObserver(*this);
+    if (!m_mainThreadPermissionObserverIdentifier)
```
This is racy since m_mainThreadPermissionObserverIdentifier gets initialized asynchronously after dispatching to the main thread and then back to the context thread. If you construct and destroy a PermissionStatus object right away, then m_mainThreadPermissionObserverIdentifier will be nullopt here, and an object will leak inside allMainThreadPermissionObservers().

cdumez replied:
One solution would be for PermissionStatus to generate m_mainThreadPermissionObserverIdentifier in its constructor and pass it to the MainThreadPermissionObserver constructor. There are likely other solutions.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
+
+        ScriptExecutionContext::postTaskTo(contextIdentifier, [weakThis, identifier = mainThreadPermissionObserverIdentifier](auto&) {
+            if (weakThis)
+                weakThis->m_mainThreadPermissionObserverIdentifier = identifier;
```
This is racy with the destructor below.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -40,7 +47,7 @@ class ScriptExecutionContext;
     WTF_MAKE_ISO_ALLOCATED(PermissionStatus);
 public:
-    static Ref<PermissionStatus> create(ScriptExecutionContext&, PermissionState, const PermissionDescriptor&);
+    static Ref<PermissionStatus> create(ScriptExecutionContext&, PermissionState, const PermissionDescriptor&, PermissionQuerySource, WeakPtr<Page>);
```
Let's not pass WeakPtr<Page> by value to avoid refcounting churn, `WeakPtr<Page>&&` is likely what we want here.

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -179,12 +179,14 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
 
-        PermissionController::shared().query(ClientOrigin { document.topOrigin().data(), WTFMove(originData) }, permissionDescriptor, document.page(), source, [contextIdentifier, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-            ScriptExecutionContext::postTaskTo(contextIdentifier, [promise = WTFMove(promise), permissionState, permissionDescriptor](auto& context) mutable {
+        auto page = source == PermissionQuerySource::DedicatedWorker ? WeakPtr { *document.page() } : nullptr;
```
So here we only initialize page if we are in a dedicated worker...

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -179,12 +179,14 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
-            ScriptExecutionContext::postTaskTo(contextIdentifier, [promise = WTFMove(promise), permissionState, permissionDescriptor](auto& context) mutable {
+        auto page = source == PermissionQuerySource::DedicatedWorker ? WeakPtr { *document.page() } : nullptr;
+
+        PermissionController::shared().query(ClientOrigin { document.topOrigin().data(), WTFMove(originData) }, permissionDescriptor, document.page(), source, [contextIdentifier, permissionDescriptor, promise = WTFMove(promise), source, page](auto permissionState) mutable {
```
... but here we pass `document.page()` unconditionally. Why the inconsistency?

RupinMittal replied:
This is a mistake. I'll update it so that if the source is a service or shared worker, then we pass in nullptr for page in both places.

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -179,12 +179,14 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
+        auto page = source == PermissionQuerySource::DedicatedWorker ? WeakPtr { *document.page() } : nullptr;
+
+        PermissionController::shared().query(ClientOrigin { document.topOrigin().data(), WTFMove(originData) }, permissionDescriptor, document.page(), source, [contextIdentifier, permissionDescriptor, promise = WTFMove(promise), source, page](auto permissionState) mutable {
+            ScriptExecutionContext::postTaskTo(contextIdentifier, [promise = WTFMove(promise), permissionState, permissionDescriptor, source, page](auto& context) mutable {
```
`page = WTFMove(page)`

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -179,12 +179,14 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
                     promise.reject(Exception { NotSupportedError, "Permissions::query does not support this API"_s });
                 else
-                    promise.resolve(PermissionStatus::create(context, *permissionState, permissionDescriptor));
+                    promise.resolve(PermissionStatus::create(context, *permissionState, permissionDescriptor, source, page));
```
WTFMove(page)

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -54,14 +61,16 @@ class PermissionStatus final : public PermissionObserver, public ActiveDOMObject
     const PermissionDescriptor& descriptor() const final { return m_descriptor; }
-    const ScriptExecutionContext* context() const final { return ActiveDOMObject::scriptExecutionContext(); }
+    PermissionQuerySource source() const final { return m_source; }
+    WeakPtr<Page> page() const final { return m_page.get(); }
```
`const WeakPtr<Page>&`

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -76,7 +85,10 @@ class PermissionStatus final : public PermissionObserver, public ActiveDOMObject
+    PermissionQuerySource m_source;
+    WeakPtr<Page> m_page;
     ClientOrigin m_origin;
+    std::optional<MainThreadPermissionObserverIdentifier> m_mainThreadPermissionObserverIdentifier;
```
std::optional<> is kinda redundant for ObjectIdentifier since ObjectIdentifier already has an invalid state and a boolean operator.

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -154,15 +154,15 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
                 promise.reject(Exception { NotSupportedError, "Permissions::query does not support this API"_s });
             else
-                promise.resolve(PermissionStatus::create(document, *permissionState, permissionDescriptor));
+                promise.resolve(PermissionStatus::create(document, *permissionState, permissionDescriptor, PermissionQuerySource::Window, WeakPtr { document->page() }));
```
`WeakPtr { }` is likely not needed here since you omitted it a few lines above.
Also, `*document->page()` since we know it is non-null and we're constructing a WeakPtr.

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -154,15 +154,15 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
 
         if (!isAllowedByFeaturePolicy(*document, permissionDescriptor.name)) {
-            promise.resolve(PermissionStatus::create(*context, PermissionState::Denied, permissionDescriptor));
+            promise.resolve(PermissionStatus::create(*context, PermissionState::Denied, permissionDescriptor, PermissionQuerySource::Window, document->page()));
```
`*document->page()` since we know it is non-null and we're constructing a WeakPtr.

szewai commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,96 @@
-    auto status = adoptRef(*new PermissionStatus(context, state, descriptor));
+    auto status = adoptRef(*new PermissionStatus(context, state, descriptor, source, page));
+    if (source != PermissionQuerySource::Window)
+        status->createAssociatedMainThreadPermissionObserver(context);
```
Why can't we do this in PermissionStatus's constructor?

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
```diff
@@ -0,0 +1,67 @@
+
+namespace WebCore {
+
+MainThreadPermissionObserver::MainThreadPermissionObserver(WeakPtr<PermissionObserver>&& permissionObserver, ScriptExecutionContextIdentifier contextIdentifier, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, WeakPtr<Page>&& page, ClientOrigin&& origin, MainThreadPermissionObserverIdentifier identifier)
```
PermissionDescriptor is just a enum, not need for `const &`

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
```diff
@@ -0,0 +1,67 @@
+{
+    m_state = newPermissionState;
+
+    ScriptExecutionContext::postTaskTo(m_contextIdentifier, [permissionObserver = m_permissionObserver, permissionState = newPermissionState](auto&) {
```
I think we should just capture `newPermissionState` here to be more concise.

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
```diff
@@ -0,0 +1,67 @@
+
+    ScriptExecutionContext::postTaskTo(m_contextIdentifier, [permissionObserver = m_permissionObserver, permissionState = newPermissionState](auto&) {
+        if (permissionObserver)
+            permissionObserver->stateChanged(permissionState);
```
And use `newPermissionState` here.

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -0,0 +1,68 @@
+    MainThreadPermissionObserver(WeakPtr<PermissionObserver>&&, ScriptExecutionContextIdentifier, PermissionState, const PermissionDescriptor&, PermissionQuerySource, WeakPtr<Page>&&, ClientOrigin&&, MainThreadPermissionObserverIdentifier);
+    ~MainThreadPermissionObserver();
+
+    MainThreadPermissionObserverIdentifier identifier() const { return m_identifier; }
```
I don't think this is needed anymore?

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -0,0 +1,68 @@
+
+class Page;
+
+class MainThreadPermissionObserver final : public PermissionObserver {
```
Note that I think the design would look a lot nicer if PermissionStatus had a MainThreadPermissionObserver no matter what (even if a main thread object) and if PermissionStatus no longer inherited PermissionObserver.

The current design is a bit awkward but this refactoring doesn't have to be done in this patch.

RupinMittal replied:
This is a good point--I just did this in a follow up patch. Now there is always a MainThreadPermssionObserver.

cdumez commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -0,0 +1,68 @@
+    PermissionState currentState() const final { return m_state; }
+    void stateChanged(PermissionState) final;
+    const ClientOrigin& origin() const final { return m_origin; }
+    const PermissionDescriptor& descriptor() const final { return m_descriptor; }
```
`const &` is not needed, this is just an enum.

cdumez commented on Source/WebCore/Modules/permissions/PermissionObserver.h:
```diff
@@ -30,20 +30,23 @@
 struct PermissionDescriptor;
 
-class PermissionObserver : public CanMakeWeakPtr<PermissionObserver> {
+class PermissionObserver : public CanMakeWeakPtr<PermissionObserver, WeakPtrFactoryInitialization::Eager> {
```
Why is this change needed? Are we ever constructing a WeakPtr to a PermissionObserver while not on the PermissionObserver's context thread? I couldn't find such case in this patch but I might have missed it.

RupinMittal replied:
I think a previous iteration of the patch was doing this, but that is no longer the case, so I'll remove this change.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,84 @@
+    return map;
+}
+
+Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, WeakPtr<Page>&& page)
```
`const &` is not needed for `PermissionDescriptor`.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,84 @@
+    if (!m_mainThreadPermissionObserverIdentifier)
+        return;
+
+    callOnMainThread([identifier = m_mainThreadPermissionObserverIdentifier]() {
```
`()` is unnecessary.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -28,10 +28,15 @@
+#include "PermissionQuerySource.h"
 #include "PermissionState.h"
+#include "ScriptExecutionContextIdentifier.h"
+#include <atomic>
```
What is this used for?

RupinMittal replied:
There is another variable that uses atomic, but I suppose this can be removed.

cdumez reviewed (approved):
LGTM

szewai commented on LayoutTests/http/tests/notifications/permission/service-worker-permission-status-onchange-event.js:
```diff
@@ -0,0 +1,11 @@
+var permissionStatus;
+
+self.onmessage = (event) => {
+  navigator.permissions.query({ name: "notifications" }).then((result)=>{
```
notifications permission only works for macOS, and http/tests/notifications/permission/ only runs on mac-wk2. Please use other permission (e.g. geolocation) for testing and put the test files in a folder that runs in different platforms.

szewai commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -179,12 +179,14 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
                     promise.reject(Exception { NotSupportedError, "Permissions::query does not support this API"_s });
                 else
-                    promise.resolve(PermissionStatus::create(context, *permissionState, permissionDescriptor));
+                    promise.resolve(PermissionStatus::create(context, *permissionState, permissionDescriptor, source, WTFMove(page)));
```
I think we prefer early return on invalid case than using `else` for valid case.

szewai commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -0,0 +1,66 @@
+    PermissionQuerySource m_source;
+    WeakPtr<Page> m_page;
+    ClientOrigin m_origin;
+    MainThreadPermissionObserverIdentifier m_identifier;
```
Where is `m_identifier` used?

RupinMittal replied:
Not use anymore, I'll remove it.

szewai commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -30,49 +30,84 @@
+#include "WorkerGlobalScope.h"
+#include "WorkerLoaderProxy.h"
+#include "WorkerThread.h"
+#include <memory>
```
Why is this needed?

RupinMittal replied:
Used for unique_ptr. But I suppose this can be removed.

szewai commented on LayoutTests/http/tests/permissions/service-worker-permission-status-onchange-event-expected.txt:
```diff
@@ -0,0 +1,15 @@
+This test checks that checks that the Permissions API on-change event works for service workers
```
"change event" or "onchange event handler"

szewai commented on LayoutTests/http/tests/permissions/service-worker-permission-status-onchange-event.html:
```diff
@@ -0,0 +1,44 @@
+<body>
+<script>
+
+description("This test checks that checks that the Permissions API on-change event works for service workers")
```
Ditto.

szewai commented on LayoutTests/http/tests/permissions/worker-permission-status-onchange-event.html:
```diff
@@ -0,0 +1,42 @@
+<body>
+<script>
+
+description("This test checks that checks that the Permissions API on-change event works for dedicated workers")
```
Ditto.

szewai commented on Tools/WebKitTestRunner/TestController.h:
```diff
@@ -588,6 +588,8 @@ class TestController {
     WebNotificationProvider m_webNotificationProvider;
     HashSet<String> m_notificationOriginsToDenyOnPrompt;
 
+    HashSet<WKRetainPtr<WKStringRef>> m_geolocationPermissionQueriers;
```
Let's use `HashSet<String> m_geolocationPermissionQueryOrigins;`
You might use `toWTFString` and `toWK` to convert it to and from WKStringRef

szewai commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -2714,6 +2731,12 @@ void TestController::decidePolicyForGeolocationPermissionRequestIfPossible()
+        return;
+
+    for (auto& originString : m_geolocationPermissionQueriers)
+        WKPagePermissionChanged(toWK("geolocation").get(), originString.get());
```
Why not invoke `WKPagePermissionChanged` in `TestController::setGeolocationPermission`?
It's bit weird that the permission request handler takes a parameter called `permissionChanged`.

RupinMittal replied:
Yeah, I initially did this because of a misinterpretation of the code. I wanted `WKPagePermissionChanged` to be invoked after the permission state gets changed locally. But that happens in `TestController::setGeolocationPermission`. So we can just invoke it after that. I'll fix this.

szewai reviewed (commented):
Almost ready; I think you need to update expectation for failed tests on bot.

szewai reviewed (approved):
LGTM

RupinMittal committed:
Add support for PermissionStatus.onchange in workers
https://bugs.webkit.org/show_bug.cgi?id=245090

Reviewed by Chris Dumez and Sihui Liu.

The onchange event of the PermissionStatus object is supposed to
be supported in both window and worker contexts. A previous patch
added support for window contexts and this patch adds support for
dedicated, service, and shared workers.

Since WebPermissionController::permissionChanged is only ever called
on the main thread, and it calls PermissionObserver::stateChanged,
the PermissionObserver it uses should be a main thread object.
Before this patch, when there was only support for window contexts,
all PermissionStatus objects were main thread objects. But now,
PermissionStatus objects can be created on worker threads as well.
When this happens, an associated MainThreadPermissionObserver is created
on the main thread and there is a one-to-one mapping between the worker
thread PermissionStatus object and the MainThreadPermissionObserver.
Then, when WebPermissionController::permissionChanged finds a
PermissionObserver to call stateChanged on, that PermissionObserver is
either a main thread PermissionStatus object (in the window case) or a
main thread MainThreadPermissionObserver, whose stateChanged function
will jump to the relevant worker thread and invoke the stateChanged
function of the associated worker thread PermissionStatus object.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/permissions/service-worker-permission-status-onchange-event-expected.txt: Added.
* LayoutTests/http/tests/permissions/service-worker-permission-status-onchange-event.html: Added.
* LayoutTests/http/tests/permissions/service-worker-permission-status-onchange-event.js: Added.
(self.onmessage):
* LayoutTests/http/tests/permissions/shared-worker-permission-status-onchange-event-expected.txt: Added.
* LayoutTests/http/tests/permissions/shared-worker-permission-status-onchange-event.html: Added.
* LayoutTests/http/tests/permissions/shared-worker-permission-status-onchange-event.js: Added.
(port.onmessage):
(onconnect):
* LayoutTests/http/tests/permissions/worker-permission-status-onchange-event-expected.txt: Added.
* LayoutTests/http/tests/permissions/worker-permission-status-onchange-event.html: Added.
* LayoutTests/http/tests/permissions/worker-permission-status-onchange-event.js: Added.
(onmessage):
* LayoutTests/imported/w3c/web-platform-tests/permissions/permissions-cg.https-expected.txt:
* LayoutTests/platform/wk2/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp: Added.
(WebCore::MainThreadPermissionObserver::MainThreadPermissionObserver):
(WebCore::MainThreadPermissionObserver::~MainThreadPermissionObserver):
(WebCore::MainThreadPermissionObserver::stateChanged):
* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h: Copied from Source/WebCore/Modules/permissions/PermissionStatus.h.
* Source/WebCore/Modules/permissions/MainThreadPermissionObserverIdentifier.h: Copied from Source/WebCore/Modules/permissions/PermissionObserver.h.
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionObserver.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::allMainThreadPermissionObservers):
(WebCore::PermissionStatus::create):
(WebCore::PermissionStatus::PermissionStatus):
(WebCore::PermissionStatus::~PermissionStatus):
(WebCore::PermissionStatus::stateChanged):
* Source/WebCore/Modules/permissions/PermissionStatus.h:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::query):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
(WebKit::WebPermissionController::permissionChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::handleQueryPermission):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::setGeolocationPermission):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: https://commits.webkit.org/254490@main
